WeakAurasOptions/OptionsFrames/TextEditor.lua

```diff
@@ -720,13 +720,57 @@ local function ConstructTextEditor(frame)
     end
   )
 
-  -- CTRL + S saves and closes
+  editor.editBox.timeMachine = {}
+  editor.editBox.timeMachinePos = 1
+  local TimeMachineMaximumRollback = 10
+
   editor.editBox:HookScript(
     "OnKeyDown",
-    function(_, key)
+    function(self, key)
+      -- CTRL + S saves and closes
       if IsControlKeyDown() and key == "S" then
         group:Close()
+      elseif key == "Z" and IsControlKeyDown() then
+        self:SetPropagateKeyboardInput(false)
+        if self.timeMachine[self.timeMachinePos + 1] then
+          self.timeMachinePos = self.timeMachinePos + 1
+          self.skipOnTextChanged = true
+          self:SetText(self.timeMachine[self.timeMachinePos][1])
+          self:SetCursorPosition(self.timeMachine[self.timeMachinePos][2])
+        end
+      elseif key == "Y" and IsControlKeyDown() then
+        self:SetPropagateKeyboardInput(false)
+        if self.timeMachine[self.timeMachinePos - 1] then
+          self.timeMachinePos = self.timeMachinePos - 1
+          self.skipOnTextChanged = true
+          self:SetText(self.timeMachine[self.timeMachinePos][1])
+          self:SetCursorPosition(self.timeMachine[self.timeMachinePos][2])
+        end
+      end
+    end
+  )
+
+  editor.editBox:HookScript(
+    "OnTextChanged",
+    function(self, userInput)
+      if not userInput then return end
+      if self.skipOnTextChanged then
+        self.skipOnTextChanged = false
+        return
+      end
+      local cursorPosition = self:GetCursorPosition()
+      local text = originalGetText(self)
+      if IndentationLib then
+        text, cursorPosition = IndentationLib.stripWowColorsWithPos(text, cursorPosition)
+      end
+      if self.timeMachine[1] and text == self.timeMachine[1][1] then
+        return
+      end
+      table.insert(self.timeMachine, 1, {text, cursorPosition - 1})
```

Comment:
There's a bug here, you need to clear the "redo" part of the history. That is if the you need to remove all entries from 1 .. self.timeMachinePos -1.

The bug can be demonstrated by entering:
ABC + Ctrl+Z
=> Text show in edit: AB
=> timeMachinePos = 2
=> timeMachine = { ABC, AB, A }

Now enter
D
=> Text show in edit: ABD
=> timeMachinePos = 1
=> timeMachine = { ABD, ABC, AB, A }
Ctrl +Z
=> Text shown in edit: ABC instead of the correct AB


I btw, would reverse the timeMachine, so that new entries are added to the end, not at the first pos. But that might be just personal preference.

Reply:
bug fixed and added 2 unrelated commits